server/webapp/WEB-INF/rails.new/app/controllers/api_v3/admin/environments_controller.rb

```diff
@@ -0,0 +1,129 @@
+##########################################################################
+# Copyright 2017 ThoughtWorks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+##########################################################################
+
+module ApiV3
+  module Admin
+    class EnvironmentsController < BaseController
+      before_action :check_admin_user_and_401
+      before_action :load_local_environment, only: [:put, :patch, :destroy]
+      before_action :check_for_stale_request, only: [:put]
+
+      def index
+        render DEFAULT_FORMAT => Admin::Environments::EnvironmentsConfigRepresenter.new(environment_config_service.getEnvironments()).to_hash(url_builder: self)
+      end
+
+      def show
+        is_query_param_provided = params.length > 3
+        is_with_remote = params[:withconfigrepo]
+        environment_name = params[:name]
+
+        if is_with_remote.nil? and !is_query_param_provided
+          load_local_environment(environment_name)
+        elsif is_with_remote and is_with_remote.downcase == 'true'
+          load_merged_environment(environment_name)
+        else
+          return render_not_found_error
+        end
+
+        json = Admin::Environments::EnvironmentConfigRepresenter.new(@environment_config).to_hash(url_builder: self)
+        render DEFAULT_FORMAT => json if stale?(etag: etag_for(environment_config_service.getEnvironmentForEdit(environment_name)))
```

Comment:
@varshavaradarajan -- we need to fetch only the editable part of the environment cause we compute the ETag only for the entity specified in `config.xml`.

`@environment_config` variable contains the merged environment whose ETag won't be same as of one from xml.